File: 0904-fruit-into-baskets/0904-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:

         mp = {}

         left = 0

         maxcnt = 0


         for right in range(len(fruits)):
            fruit_type = fruits[right]

            mp[fruit_type] = mp.get(fruit_type , 0) + 1

            while len(mp)>2:
                removable_fruit_type = fruits[left]
                mp[removable_fruit_type]-=1
                if mp[removable_fruit_type] == 0:
                     del mp[removable_fruit_type]
                
                left+=1
                
            
 
        

            maxcnt = max(right - left + 1 , maxcnt)





         return maxcnt
        

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.js
/**
 * @param {number[]} fruits
 * @return {number}
 */
var totalFruit = function (fruits) {

    const map = new Map();

    let n = fruits.length

    let left = 0;

    let right = 0;

    let max = 0;


    for (let right = 0; right < n; right++) {

        let fruitType = fruits[right];

        map.set(fruitType, (map.get(fruitType) || 0) + 1)

        while (map.size > 2) {
            let leftFruitType = fruits[left]


            map.set(leftFruitType, map.get(leftFruitType) - 1);
            if (map.get(leftFruitType) === 0) map.delete(leftFruitType)

            left++

        }

        max = Math.max(max, right - left + 1)






    }


    return max;

};